### DISCUSSION TITLE
浏览器扩展版本的进度 🚀

### DISCUSSION THREAD

**Discussion by beijixiaohu** [beijixiaohu/OJBetter#173] [Show and tell]
该脚本即将推出浏览器扩展（又称浏览器插件）版本✨

**🔧 规划：**

- 计划分为两个扩展，一个专注于任意网站的翻译和汉化，名称待定；一个整合当前 OJ Better 各个插件的小功能，名称仍然为 OJ Better。
- 其中翻译扩展将会仅提供浏览器扩展版本，而 OJ Better 扩展将同时提供篡改猴（Tampermonkey）版本。

**🔒 新的扩展将：**

- 继续开源，届时欢迎贡献代码。
- 采用现代前端开发的主流工具链，这将使代码更易读、易于协作和扩展，并更容易实现跨浏览器兼容。

**📜 关于旧的脚本：**

- 新的扩展发布后，旧的脚本将仅进行维护，不再添加新功能，脚本会一直保留，敬请放心。

此外，作者会尽力维持良好的性能，至少不低于对应功能在市面上已有的商用扩展（如果有）。这相当复杂，当前已经顺利完成了一部分核心功能，预计还需数月的开发，敬请期待！✨

**💬 作者会在讨论区中分享当前的进展和开发心得，欢迎有兴趣的朋友前往 #173 讨论。**

**Comment by beijixiaohu** [comment]
**🎉 网页全文翻译目前已经实现了主要功能**

先说说目前的主要实现思路：

**获取文本：**

- 使用原生的 `innerText` 属性获取元素的可读文本，其排版接近实际看到的效果。
- 根据 `innerText` 返回的字符数，将页面拆分为若干个大块的元素区域，每块字符数不超过指定值。
- 对每块区域，将文本按照换行符和制表符分割成条，保存到 `pageSegmentationMap: Map<HTMLElement, string[]>`，因为 HTML 本身就是结构化的，因此每块内的字符是语义相关的，一起发送给翻译服务器可确保上下文，从而获得更好的翻译结果。

**译文附着：**

- 为每条文本添加唯一 ID 进行关联，使用双向映射 `idToOriginalMap` 方便根据原文查找id。
- 递归每个区域内的所有元素，判断每个元素的 `innerText` 是否在 `idToOriginalMap` 内。
- 若在，则递归查找可维持 `innerText` 不变的最小元素，创建容器插入到其内部后面（若为文本节点，则插入到外部后面）。
- <del>为实现样式隔离，以及减少对依赖网页上的原文文本来执行特定逻辑的 JS （通常是一些脚本）的影响，容器内使用 `shadow dom`。使用 `idToShadowRootMap` 保存各原文对应的 `shadow root`，便于后续响应式的实现。</del>放弃了使用 shadow dom ，改为使用普通的 span 标签，因为会导致”元素保留“体验不佳，样式隔离使得译文中的保留的元素在shadow dom中显示异常。
- 此外，为保证尽量不破坏页面原本样式，会判断当前翻译结果是否导致了换行，如果产生了换行，则直接另起一行显示翻译结果，否则不换行显示，这样会更美观

**翻译：**

- 将每块的原文发送给翻译服务器，收到译文后将原文与译文关联，存入 `idToTranslatedMap`。

后文跟帖说，评论区因为过长会被折叠，点击下面蓝色的 `Show  previous reply` 即可展开

**Reply by beijixiaohu** [reply]
下面来说说翻译功能的响应式设计：

翻译相关的各个 Map 统一放在 `useEntireSiteTranslatorStore.ts` 中，使用 Pinia 管理状态。

对于每一个网页，会创建且仅创建一个 Vue 实例。页面上的所有翻译结果通过模板动态渲染，并使用 Teleport 传送到对应的 Shadow Root。

这样可以轻松实现状态更新，例如从翻译中到翻译结束的转换，以及使用 GPT 翻译时的流式传输动态更新。

利用响应式，扩展还实现了翻译网页的各个步骤间依赖关系的分离：

基于发布订阅模式，将 "页面分割"、"插入容器"、"翻译" 这三个步骤相互独立，

在 `useEntireSiteTranslatorStore.ts` 中使用 rxjs 创建两个 Subject： `pageSegmentationSubject`、`segmentToDataSubject`

- 页面分割模块 在分割后通过 `pageSegmentationSubject` 广播。
- 插入容器模块 订阅 `pageSegmentationSubject`，在创建文本的唯一 ID 并插入容器后，通过 `segmentToDataSubject` 广播。
- 翻译模块 订阅 `segmentToDataSubject`，一旦有新文本，就进行翻译并将结果存入 `idToTranslatedMap`。

**Reply by beijixiaohu** [reply]
再说说所做的性能优化：

**任务队列：**

扩展中利用 `requestAnimationFrame` 实现了一个基于屏幕刷新率的轮转任务队列，将页面分割和插入容器这两个长递归任务进行队列管理。

这得益于上述的依赖关系分离，通过轮转可以让"插入容器"任务更快的进行，而无需等待页面完全分割完成，让用户更早看到部分翻译结果。

此外，通过统计任务耗时，限制一次动画时间内的任务总耗时不超过 16ms，这样可以避免线程长时间阻塞，让交互事件可以正常响应。这也大幅缩短了渲染和绘制的总耗时。

**缓存重计算：**

由于 innerText 和元素可见性、换行检查等操作会导致页面重排重绘，是昂贵的计算，因此使用 Pinia 缓存，减少重排重绘的发生次数。

还有一些细节优化，作者一时想不起来了，有的话后面再补充

**Reply by beijixiaohu** [reply]
与当前流行的两个翻译插件的实测对比截图，使用相同的翻译服务（DeepL）：

测试网页链接：[https://github.com/isaurssaurav/hot-reload-extension-vite-plugin/issues](https://github.com/isaurssaurav/hot-reload-extension-vite-plugin/issues)

**OJBetter：** （全文翻译）

![image-20240819125620611](https://github.com/user-attachments/assets/93af1de1-616a-4e39-8719-67147d468cb1)

耗时（性能测试为多次测试取中，下同）：

![image](https://github.com/user-attachments/assets/238e3c5c-009b-4496-927a-3c3a5f86c7e1)

**沉浸式翻译：** （全文翻译）

![image-20240819125517642](https://github.com/user-attachments/assets/49723219-3213-4724-8c85-9d9d6d72210e)

耗时：

![image](https://github.com/user-attachments/assets/f509bec2-4452-4e86-a58c-658907571985)

**划词翻译：** （其不能直接全文翻译，只能一次只翻译可视区域，由于网页很短，因此也相当于全文翻译）

![image](https://github.com/user-attachments/assets/cf32ce73-2d3f-419b-b3fd-307a67a19efb)

耗时：

![image](https://github.com/user-attachments/assets/b1c7c542-5cd0-40d0-a77e-74765bbdad1f)

**Reply by beijixiaohu** [reply]
再贴一个，同样使用相同的翻译服务（DeepL），性能测试均进行多次，截图为其中的中间值：

测试网页链接：[https://netflixtechblog.com/java-21-virtual-threads-dude-wheres-my-lock-3052540e231d](https://netflixtechblog.com/java-21-virtual-threads-dude-wheres-my-lock-3052540e231d)

**OJBetter:** 全文翻译

<img width="1015" alt="Snipaste_2024-08-19_16-26-28" src="https://github.com/user-attachments/assets/1b2612f9-f413-4954-9a64-b8512cc2bb17">

耗时：

![image](https://github.com/user-attachments/assets/2c7773f2-ae2d-4072-b011-7abe0e2f3bd1)

**沉浸式翻译：** 全文翻译

![image](https://github.com/user-attachments/assets/7a1d6708-4989-4994-a928-c2e165f73691)

耗时：

![image](https://github.com/user-attachments/assets/9caa4a0e-feee-4012-9cc0-1f4fefe42f9c)

**划词翻译：** 非全文翻译

![image](https://github.com/user-attachments/assets/5cd25b06-0b2a-4ab2-9f4f-ed04c654fe8a)

其不支持立即翻译全文，手动滚动让其翻译全文所得的耗时干扰因素太多，没有意义，因此不贴出其耗时截图

**Reply by beijixiaohu** [reply]
可以看到，总的来说，OJBetter的性能还是不错的，不会影响正常的网页浏览。

而且由于实现方式不同，这三个插件即使使用相同的翻译服务，得到的结果也有不少差别，

世上安得两全法，而且当前的方案也有很多改进的空间，如果你有更好的想法或者建议，欢迎分享🌹

**Reply by beijixiaohu** [reply]
值得一提的是，OJBetter 目前还未对任何网站做特殊优化，但是世上网站千千万，还是有很多布局比较奇葩但是又很常用的网站的，为此仍然需要对一些网站进行适配，（比如 medium 的文章底部推荐部分。。。）

关于如何更好的对特殊网站进行适配作者还没有想好，后续有进展了再来分享，感谢阅读

**Comment by ZZC2012** [comment]
把插件转给我呗，不是上次转我的那个啥也没有的插件

**Reply by beijixiaohu** [reply]
待功能相对完善后会提供测试版本

**Reply by ZZC2012** [reply]
OK

**Reply by ZZC2012** [reply]
不过你上次那个......真的是一言难尽呀。

**Comment by beijixiaohu** [comment]
**🎉 现已实现页面中动态元素的翻译**

下面分享下这两周所做的工作，

由于之前的网页全文翻译实现方式存在一些难以处理的问题：

- 无法区分上下文中的两段相同的文本，导致译文显示不全
- 难以实现页面动态元素的翻译，因为元素变动导致的文本变动难以追踪

经过了一段时间的研究，最终**放弃了之前的原文获取的实现方式**，改为了如下：

**原文获取：**

对于每一个切分块，不再如之前那样处理，而是将区块划分为一个个的翻译单元 `translationUnit`：

从块的根节点开始检查每个子节点。

- 如果节点是文本且可见，就积累到暂存列表。
- 如果遇到内联元素，继续递归检查其子节点。
- 如果遇到块级元素，就把当前积累的文本节点组合成一个翻译单元，然后清空暂存列表，继续递归处理子节点。
- 如果遇到表格，特殊处理：遍历每个单元格，把它们的内容提取出来，并为每个单元格生成一个翻译单元。

对于每一个翻译单元，使用 `useNodeToIdMapStore` 维护唯一 ID，方便追踪。

**译文附着：**

由于 `translationUnit` 中包含节点们的引用，因此添加译文容器非常简单，只需要添加到最后一个节点的后面即可，其他的基本与之前相同。

对了，上述词条获取的方式也带来了新的问题：

如果最后一个元素是包裹在 `a`、`sub` 、`sup` 内的，则可能导致大段地译文都被错误的包裹在其中，不美观且难以阅读（比如很多维基百科的页面）

这里目前是这样处理的：如果译文文本超过 20 字符，则认为被 `a`、`sub` 、`sup` 包裹是不易阅读的，

会向上查找祖先元素，直到找到不会被包裹的，将容器添加到其的后面。

**Reply by beijixiaohu** [reply]
还调整了翻译数据状态管理相关的流程，

去掉了 `pageSegmentationMap` 和对应的 subject，因为发现并不需要额外维护它们，都放在 `segmentToData` 中就行，

也对翻译数据更新逻辑进行了调整，这里就不啰嗦了

**Reply by beijixiaohu** [reply]
现在的实测截图（理论上来说应该性能会好很多，实际上好像变化不太大哈哈哈）：

测试网页链接：https://github.com/isaurssaurav/hot-reload-extension-vite-plugin/issues

<img width="945" alt="Snipaste_2024-08-31_23-29-56" src="https://github.com/user-attachments/assets/2e4faa53-99bf-446b-bf0c-81277fd1bf5f">
<img width="152" alt="Snipaste_2024-08-31_23-30-10" src="https://github.com/user-attachments/assets/3381bc84-adeb-4b58-9970-d1cbbce235e2">

测试网页链接：https://netflixtechblog.com/java-21-virtual-threads-dude-wheres-my-lock-3052540e231d

<img width="933" alt="Snipaste_2024-08-31_23-30-23" src="https://github.com/user-attachments/assets/0d96f1ef-3f13-479c-8b40-f85ef94f75ce">
<img width="119" alt="Snipaste_2024-08-31_23-30-32" src="https://github.com/user-attachments/assets/aa83eac5-7e7c-4bed-8ec4-657aeeaee9eb">

**Comment by ghost** [comment]
能不能加入一个 html to markdown with katex 的东西，有很多古老的题解翻译不了（12000字）

**Reply by beijixiaohu** [reply]
会有的

**Comment by beijixiaohu** [comment]
**🎉 现已实现翻译任务中的一些关键模块**

包括：缓存、限流、缓冲

**缓存：**

每一次翻译后，会将 翻译文本的 sha256 值 和 压缩的翻译结果数据（使用 lz-string ）存放在扩展的 indexdb 数据库中，

对于 sha256 相同的翻译文本，再次遇到时会直接从数据库中取数据，而不必再重复翻译。

**限流：**

使用令牌桶算法来限流，对于每个翻译服务，均可以单独设置其令牌池大小和令牌填充速率。

**缓冲：**

使用 RxJS 来合并小的翻译请求。

创建一个 `translateSubject` 来收集任务，每次收到新任务时，同时加入缓冲区 `taskBuffer`。

监听 ”定时器 `timerBuffer`“ 和 ”任务流 `taskSubject`“，当缓冲区中的字符数超过设置值或者时间到了时，缓冲区中的所有任务会被批量合并翻译。

**Comment by beijixiaohu** [comment]
**🎉 本周进展**

**实现了译文样式切换和样式自定义**

![image](https://github.com/user-attachments/assets/4cee185a-a544-40cb-87ef-097269d192f7)
---
![image](https://github.com/user-attachments/assets/3a352d77-02c3-47e6-a687-686a90152e49)
---
![image](https://github.com/user-attachments/assets/ed6ceb1f-4eb0-4f68-b4e5-70d6b129965e)

**实现了"忽略与保持"相关的功能和自定义配置**

![image](https://github.com/user-attachments/assets/6bafc8b1-3ec7-4d91-9f67-476efd39c3f0)

**Reply by beijixiaohu** [reply]
**实测截图：**

https://codeforces.com/contest/2009/problem/A
![image](https://github.com/user-attachments/assets/6860d27b-1bfc-4f93-9985-8ee3c60fa4ee)
https://codeforces.com/contest/167/problem/A
![image](https://github.com/user-attachments/assets/e8c97873-c75d-4ea1-8c4e-53b03feeded2)
https://codeforces.com/blog/entry/4214
![image](https://github.com/user-attachments/assets/cc7cf080-04fa-421d-8e4b-42d02741c89b)

**Comment by ZZC2012** [comment]
加油，等插件相对完整时发在这里

**Comment by beijixiaohu** [comment]
**🎉 本周进展**

区域翻译

![image](https://github.com/user-attachments/assets/28b09034-9239-4fbf-8d73-10eb6c7d679a)
![image](https://github.com/user-attachments/assets/fad4ed21-28ec-4e6d-af66-545db704e571)
![image](https://github.com/user-attachments/assets/63347181-cdfb-4a97-a4f6-2e5af952b13f)

**Reply by beijixiaohu** [reply]
效果截图：

![image](https://github.com/user-attachments/assets/88715f53-9434-4957-ac5f-1aa3459d5ba1)
![image](https://github.com/user-attachments/assets/14fce098-505b-4da9-8d83-2badec3312ae)
![image](https://github.com/user-attachments/assets/a2699c6a-8eaf-4483-8aae-ed36fb6d67a2)

**Comment by beijixiaohu** [comment]
**🎉 本周进展**

现已部分实现网站本地化功能：

![image](https://github.com/user-attachments/assets/c02997fd-fab1-42ea-905a-234f2ebbd4ba)
![image](https://github.com/user-attachments/assets/37228cc2-8330-4b79-981b-77f25000ee2e)
![image](https://github.com/user-attachments/assets/13293d8f-898b-4d45-92b0-4fea1db828a8)
![image](https://github.com/user-attachments/assets/0a5a04ea-aa8f-45bf-87ae-d15070fe64da)
![image](https://github.com/user-attachments/assets/8e1163d3-0347-4ec3-b821-b6148dd12cda)

![image](https://github.com/user-attachments/assets/9aae329c-79a8-481c-bc15-06c312996bf7)
![image](https://github.com/user-attachments/assets/f51f2ac0-7963-41a3-9278-7affc1538e80)

**Reply by beijixiaohu** [reply]
这一部分的实现非常复杂，预计还需三周以上的时间来完成

**Comment by wrk-123** [comment]
应该有批量添加/导出规则？

**Reply by beijixiaohu** [reply]
确实，此外后期计划提供一个后端平台，来支持协作，同步规则

**Comment by bakebakebakebake** [comment]
加油，我在等着用ipad打CF呢😀👍

**Reply by beijixiaohu** [reply]
感谢支持，但是ipad的支持优先级不会很高，因为之前搜了下，ipad的浏览器扩展要想上架商店，需要一年交99美元的开发者认证……

![image](https://github.com/user-attachments/assets/a1c4e466-982d-4573-8786-853f518256a8)

因此除非某一天起，扩展开始能获得正向收益，否则不会考虑

**Reply by bakebakebakebake** [reply]
原来是这样的么，那若上架浏览器拓展商店的话可能是得需要收费的🥲，如果是其他浏览器(非safari但似乎还是safari内核)的话呢可不可以支持这个插件呢

**Reply by beijixiaohu** [reply]
不行，ios系统上的浏览器都是safari套壳
![image](https://github.com/user-attachments/assets/da8c1e49-797f-43ea-bb28-eab5db98dd71)
mac上的chrome倒是可以

**Comment by ZZC2012** [comment]
发一下呗

**Reply by beijixiaohu** [reply]
做好了会发的，别着急

**Reply by ZZC2012** [reply]
所以你给出那么多截图，让人很想要一个！！

**Comment by wrk-123** [comment]
现在能不能提一下进展

**Reply by beijixiaohu** [reply]
感谢关注，这段时间很忙，所以项目整体进展不大，基本属于每天新增几行代码的情况。

而且现在需要实现的几个功能都是不好实现的那种，特别是”译文原文对照“这个功能，卡了很久，远比之前我想的要复杂，等这个实现了会再来这里分享一下思路。

**Reply by wrk-123** [reply]
> 感谢关注，这段时间很忙，所以项目整体进展不大，基本属于每天新增几行代码的情况。
> 
> 而且现在需要实现的几个功能都是不好实现的那种，特别是”译文原文对照“这个功能，卡了很久，远比之前我想的要复杂，等这个实现了会再来这里分享一下思路。

如果实在太难的话可以先放一下

**Reply by beijixiaohu** [reply]
> 如果实在太难的话可以先放一下

这个功能已经基本实现啦

https://github.com/user-attachments/assets/0a1a1348-130b-4a58-92a4-d03103fe0aca

**Comment by wrk-123** [comment]
新年好，加油

**Reply by beijixiaohu** [reply]
谢谢，新年好❤️🎇

**Reply by wrk-123** [reply]
> 谢谢，新年好❤️🎇
话说浏览器扩展咋样了，挺期待且好奇的

**Reply by beijixiaohu** [reply]
目前进度在50%左右，还有很多工作没有完成，尤其是后端部分，我录几个gif展示下

**Reply by beijixiaohu** [reply]
这是网站汉化的“快捷添加规则“功能

https://github.com/user-attachments/assets/2413b222-a768-4a27-8119-b8c276020ab8

**Reply by beijixiaohu** [reply]
这是单击扩展图标的弹出页以及单词本

https://github.com/user-attachments/assets/dde1760f-69b5-46f7-994a-899b108c5edc

**Reply by beijixiaohu** [reply]
划词


https://github.com/user-attachments/assets/43202b79-048a-4093-b3bd-8514ef044443

**Reply by beijixiaohu** [reply]
目前具体的功能进展就这些了，感谢关注

**Comment by qwedc001** [comment]
有个期望上架时间嘛，关注浏览器扩展版本挺久的了）

**Reply by beijixiaohu** [reply]
感谢关注，现在70%左右，还得几个月吧，具体时间我也说不准

**Comment by wrk-123** [comment]
1月份的时候提到：进度在 50% 左右。那现在？关心一下

**Reply by beijixiaohu** [reply]
70%吧，隔几天写一点，还早还早

**Reply by cooluoCN** [reply]
> 70%吧，隔几天写一点，还早还早

现在不早喽 😢

**Comment by lzm0107** [comment]
现在怎么样了哇

**Reply by beijixiaohu** [reply]
感谢关注，不过说来惭愧，目前进度大概在80%……

不是因为我在偷懒，而是因为我低估了我想实现的功能的复杂程度哈哈哈哈

最近专注在后端，基本每天都有在写，刚查了下后端的git记录，看图

<img width="865" height="388" alt="image" src="https://github.com/user-attachments/assets/2d2278eb-10a1-4fea-9b79-925238ccbf56" />

预估到元旦左右就能发布第一个测试版本了

**Reply by wrk-123** [reply]
> 感谢关注，不过说来惭愧，目前进度大概在80%……
> 
> 不是因为我在偷懒，而是因为我低估了我想实现的功能的复杂程度哈哈哈哈
> 
> 最近专注在后端，基本每天都有在写，刚查了下后端的git记录，看图
> 
> <img alt="image" width="865" height="388" src="https://private-user-images.githubusercontent.com/60805870/498337630-2d2278eb-10a1-4fea-9b79-925238ccbf56.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc1MTIwNTIsIm5iZiI6MTc2NzUxMTc1MiwicGF0aCI6Ii82MDgwNTg3MC80OTgzMzc2MzAtMmQyMjc4ZWItMTBhMS00ZmVhLTliNzktOTI1MjM4Y2NiZjU2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTA0VDA3MjkxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRiMmY1OGM2ZjgyYjBmN2I5ODEzNGMyYWE5ZjI0NWE1M2ZkODVhMGFhYWE5ZWM4Mjc2OGVjYzk5Y2MwNjRhN2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.icZBY231pRfWFZSTokcDQchhgSSy3IbdU0qIbsz_yUU">
> 预估到元旦左右就能发布第一个测试版本了

元旦左右，是现在吗

**Reply by cooluoCN** [reply]
> 感谢关注，不过说来惭愧，目前进度大概在80%……
> 
> 不是因为我在偷懒，而是因为我低估了我想实现的功能的复杂程度哈哈哈哈
> 
> 最近专注在后端，基本每天都有在写，刚查了下后端的git记录，看图
> 
> <img alt="image" width="865" height="388" src="https://private-user-images.githubusercontent.com/60805870/498337630-2d2278eb-10a1-4fea-9b79-925238ccbf56.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc1MzA1MTIsIm5iZiI6MTc2NzUzMDIxMiwicGF0aCI6Ii82MDgwNTg3MC80OTgzMzc2MzAtMmQyMjc4ZWItMTBhMS00ZmVhLTliNzktOTI1MjM4Y2NiZjU2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTA0VDEyMzY1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJmOTIxOTk0MTA4OTUxZGZlZTg4NTJmNWY2ZTBlNTZmMDhhZmMzODRhZGZhZWEzYTJlYTc0ZmM1MzFkNzE3ZmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.eINvQZnRgdWd8rRVeMnaYsFkL0aSBqgHeQETGzgIFYk">
> 预估到元旦左右就能发布第一个测试版本了

小狐预估失败 T_T 话说去年一月的演示看起来已经很不错了，现在情况怎么样了呢

**Reply by beijixiaohu** [reply]
@cooluoCN @wrk-123 感谢兄弟们的等待，~目前仍在稳步推进~难点已经基本解决，但还有大量细节需要完善，仍需一段时间。

翻了下之前的截图，感慨良多，现在的界面与之前的大不相同，一年多过去了，别的不说，审美这一块确实涨了一些😂。

附几张现在的插件端的截图：

弹出页面

<img width="1430" height="1119" alt="image" src="https://github.com/user-attachments/assets/b92f2d67-97b3-42af-9ed3-73843e8be910" />

网站汉化页面

<img width="1429" height="1118" alt="image" src="https://github.com/user-attachments/assets/910cbdbe-8398-44b0-913d-13110aeab208" />

<img width="1429" height="1118" alt="image" src="https://github.com/user-attachments/assets/1d8a2aad-9445-41bc-b8db-6f6b29e8292d" />

还有目前仍然只是个壳子的翻译工作流

<img width="1429" height="1119" alt="image" src="https://github.com/user-attachments/assets/78337f49-77f5-423b-ace3-c705dc29206a" />

还有更多的时间是付出在云平台上的，就暂不细说了，如有更多进展，再来这里给各位分享。

另祝各位2026新年快乐~❤️

**Reply by cooluoCN** [reply]
新年快乐 ❤️ 一大进步之有自己的 logo 和网站(?)了 😆

**Reply by qwedc001** [reply]
新年快乐❤️

从选手开始期待到现在已经退役了😆只能看着小登们享受了

[a comment by an automated account is not shown]
